src/sage/modular/drinfeld_modform/ring.py

```diff
@@ -0,0 +1,726 @@
+                                 f"(={self.rank()})")
+            return self._generator_coefficient_form(i)
+        try:
+            K = self._base_ring
+            T = K.gen()
+            a = K(a)
+        except TypeError:
+            raise TypeError("a should be an element of the base ring")
+        if a.valuation(T) < 0:
+            raise ValueError("a should be in the ring of regular functions")
+        a = a.numerator()
+        if i < 1 or i > a.degree()*self.rank():
+            raise ValueError(f"index (={i}) must be >= 1 and <= deg(a)*rank "
+                             f"(={a.degree()*self.rank()})")
+        coeff_forms = self._coefficient_forms(a)
+        return coeff_forms[i - 1]
+
+    def coefficient_forms(self, a=None):
+        r"""
+        Return the list of all coefficients of the universal Drinfeld
+        module at `a`.
+
+        See also :meth:`coefficient_form` for definitions.
+
+        INPUT:
+
+        - ``a`` -- (default: ``None``) an element in the ring of regular
+          functions. If `a` is ``None``, then the method returns the
+          coefficients forms at `a = T`.
+
+        OUTPUT: a list of Drinfeld modular forms. The `i`-th element of
+        that list corresponds to the `(i+1)`-th coefficient form at `a`.
+
+        EXAMPLES::
+
+            sage: q = 3
+            sage: A = GF(q)['T']
+            sage: K.<T> = Frac(A)
+            sage: M = DrinfeldModularForms(K, 2)
+            sage: M.coefficient_forms()
+            [g1, g2]
+            sage: M.coefficient_forms(T^2)
+            [(T^3 + T)*g1, g1^4 + (T^9 + T)*g2, g1^9*g2 + g1*g2^3, g2^10]
+            sage: M.coefficient_forms(T^3)
+            [(T^6 + T^4 + T^2)*g1,
+             (T^9 + T^3 + T)*g1^4 + (T^18 + T^10 + T^2)*g2,
+             g1^13 + (T^27 + T^9 + T)*g1^9*g2 + (T^27 + T^3 + T)*g1*g2^3,
+             g1^36*g2 + g1^28*g2^3 + g1^4*g2^9 + (T^81 + T^9 + T)*g2^10,
+             g1^81*g2^10 + g1^9*g2^28 + g1*g2^30,
+             g2^91]
+
+        TESTS::
+
+            sage: q = 3
+            sage: A = GF(q)['T']
+            sage: K.<T> = Frac(A)
+            sage: M = DrinfeldModularForms(K, 2)
+            sage: M.coefficient_forms(1/T)
+            Traceback (most recent call last):
+            ...
+            ValueError: a should be in the ring of regular functions
+            sage: M.coefficient_forms(x)
+            Traceback (most recent call last):
+            ...
+            TypeError: a should be an element of the base ring
+        """
+        K = self._base_ring
+        T = K.gen()
+        if a is None:
+            return [self._generator_coefficient_form(i)
+                    for i in range(1, self.rank() + 1)]
+        try:
+            a = K(a)
+        except TypeError:
+            raise TypeError("a should be an element of the base ring")
+        if a.valuation(T) < 0:
+            raise ValueError("a should be in the ring of regular functions")
```

Comment:
I'm not sure this is correct.
I think you want to avoid also e.g. $\frac 1{T+1}$, don't you?

Reply:
Indeed, I will fix it!